plugins/catalog-backend/src/database/types.ts

```diff
@@ -153,7 +153,10 @@ export type Database = {
     matchingGeneration?: number,
   ): Promise<DbEntityResponse>;
 
-  entities(tx: unknown, filters?: EntityFilters): Promise<DbEntityResponse[]>;
+  entities(
+    tx: unknown,
+    filters?: EntityFilters | EntityFilters[],
```

Comment:
This is more or less an internal interface that doesn't need to be super easy to type. I'd go with only `EntityFilters[]` instead.

Reply:
Sure.

I also thought there's a little too much logic spread around, and EntityFilters could be an interface that exposes `matchNulls`, `matchIn` etc. An implementing class of that could have static builder methods that deal with the parsing of the query string format, or translating from the object format. That would concentrate all of this better and remove a fair bit of code from the db layer.